omp/test/distributed/matrix_kernels.cpp

```diff
@@ -0,0 +1,384 @@
+                                                                 mapping,
+                                                                 num_parts);
+
+    this->validate(
+        partition.get(), partition.get(), d_partition.get(), d_partition.get(),
+        gko::device_matrix_data<value_type, global_index_type>{this->ref});
+}
+
+
+TYPED_TEST(Matrix, BuildsLocalSmallIsEquivalentToRef)
+{
+    using value_type = typename TestFixture::value_type;
+    using local_index_type = typename TestFixture::local_index_type;
+    using global_index_type = typename TestFixture::global_index_type;
+    gko::distributed::comm_index_type num_parts = 3;
+    gko::size_type num_rows = 10;
+    gko::size_type num_cols = 10;
+    auto mapping =
+        gko::test::generate_random_array<gko::distributed::comm_index_type>(
+            num_rows,
+            std::uniform_int_distribution<gko::distributed::comm_index_type>(
+                0, num_parts - 1),
+            this->engine, this->ref);
+    auto input = gko::test::generate_random_device_matrix_data<
+        value_type, global_index_type>(
+        num_rows, num_cols,
+        std::uniform_int_distribution<int>(0, static_cast<int>(num_cols - 1)),
+        std::uniform_real_distribution<gko::remove_complex<value_type>>(0, 1),
+        this->engine, this->ref);
+
+    auto partition = gko::distributed::Partition<
+        local_index_type, global_index_type>::build_from_mapping(this->ref,
+                                                                 mapping,
+                                                                 num_parts);
+    auto d_partition = gko::distributed::Partition<
+        local_index_type, global_index_type>::build_from_mapping(this->exec,
+                                                                 mapping,
+                                                                 num_parts);
+
+    this->validate(partition.get(), partition.get(), d_partition.get(),
+                   d_partition.get(), input);
+}
+
+
+TYPED_TEST(Matrix, BuildsLocalIsEquivalentToRef)
+{
+    using value_type = typename TestFixture::value_type;
+    using local_index_type = typename TestFixture::local_index_type;
+    using global_index_type = typename TestFixture::global_index_type;
+    gko::distributed::comm_index_type num_parts = 13;
+    gko::size_type num_rows = 67;
+    gko::size_type num_cols = 67;
+    auto mapping =
+        gko::test::generate_random_array<gko::distributed::comm_index_type>(
+            num_rows,
+            std::uniform_int_distribution<gko::distributed::comm_index_type>(
+                0, num_parts - 1),
+            this->engine, this->ref);
+    auto input = gko::test::generate_random_device_matrix_data<
+        value_type, global_index_type>(
+        num_rows, num_cols,
+        std::uniform_int_distribution<int>(static_cast<int>(num_cols),
+                                           static_cast<int>(num_cols)),
```

Comment:
I might have wanted to ensure that no parts are empty, so I used a dense matrix here. I don't think it would make any difference though. I will keep it as it is.